packages/yew/src/virtual_dom/key.rs

```diff
@@ -1,74 +1,189 @@
 //! This module contains the implementation yew's virtual nodes' keys.
 
 use crate::html::ImplicitClone;
-use std::fmt::{self, Display, Formatter};
-use std::ops::Deref;
+use std::hash::Hash;
 use std::rc::Rc;
+use std::sync::Arc;
+
+// Invariants that must hold for the key implementation to work since `usize`
+// and `isize` are casted to `u128` and `i128`.
+// TODO: Uncomment these when const assertions are available in MSRV.
+// const _: () = assert!(std::mem::size_of::<usize>() <= std::mem::size_of::<u128>());
+// const _: () = assert!(std::mem::size_of::<isize>() <= std::mem::size_of::<i128>());
+
+// Some seeds / salts picked at random for types which should be treated
+// uniformly but distinct from each other.
+const UNSIGNED_SEED: u64 = 0x3a85f087a0fd37aa;
+const SIGNED_SEED: u64 = 0x607899f4fe4b09fa;
+const BYTES_SEED: u64 = 0x27422aa0116c109c;
+
+#[cfg(feature = "twox-hash")]
+mod imp {
+    use std::hash::Hash;
+    use twox_hash::xxh3::{Hash128, HasherExt};
+
+    pub(super) type Repr = u128;
+
+    #[inline]
+    pub(super) fn from_hash<T>(seed: u64, value: T) -> Repr
+    where
+        T: Hash,
+    {
+        let mut hasher = Hash128::with_seed(seed);
+        value.hash(&mut hasher);
+        hasher.finish_ext()
+    }
+}
+
+#[cfg(not(feature = "twox-hash"))]
+mod imp {
+    use std::collections::hash_map::DefaultHasher;
+    use std::hash::{Hash, Hasher};
+
+    pub(super) type Repr = u64;
+
+    #[inline]
+    pub(super) fn from_hash<T>(seed: u64, value: T) -> Repr
+    where
+        T: Hash,
+    {
+        let mut hasher = DefaultHasher::new();
+        seed.hash(&mut hasher);
+        value.hash(&mut hasher);
+        hasher.finish()
+    }
+}
 
 /// Represents the (optional) key of Yew's virtual nodes.
 ///
 /// Keys are cheap to clone.
-#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq, Hash)]
-pub struct Key {
-    key: Rc<str>,
-}
+#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq, Hash)]
+#[repr(transparent)]
+pub struct Key(self::imp::Repr);
 
-impl Display for Key {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        self.key.fmt(f)
+// Default implementation for byte sequences.
+impl From<&[u8]> for Key {
```

Comment:
Why not `impl<T> From<T> for Key where T: Hash` and only specialise on types that needs special handling?

Reply:
Providing a blanket `impl<T> From<T> for Key where T: Hash` would prevent any specialization, since `Hash` might be implemented upstream for any `T` in the future (trait coherence rules) and we don't have access to stable `specialization.

Reply:
I guess instead of using `From<T>`, having `pub fn new<H: Hash>(h: H) -> Key` as an associated function would work.

Usage of `From`(Into) for conversion into Keys is designated by the `html!` macro and can be changed to something else.

Reply:
I don't see how it semantically makes sense to construct a `Key` from an abstract hashable type. I see the hashing as a performance improvement that acts as-if it would compare the actual values the `Key` was constructed from. I wouldn't expand this. Users can just hash those types themselves before constructing the keys and wrestle their own collision problems.